File: mcp-client-bridge/src/http-streamable-client.ts
/**
 * HTTP Streamable client — POST /mcp with JSON-RPC.
 * Simple stateless request-response. No SSE, no sessions.
 */

import { BridgeConfig } from './bridge-config.js';

interface JsonRpcResponse {
  jsonrpc: '2.0';
  id: number | null;
  result?: unknown;
  error?: { code: number; message: string };
}

interface ToolCallResult {
  content?: Array<{ type: string; text: string }>;
  isError?: boolean;
}

export class HttpStreamableClient {
  private requestIdCounter = 0;
  private _connected = false;

  constructor(private readonly config: BridgeConfig) {}

  get isConnected(): boolean {
    return this._connected;
  }

  async initialize(): Promise<boolean> {
    try {
      const response = await this.post('initialize', {
        protocolVersion: '2025-03-26',
        capabilities: {},
        clientInfo: {
          name: 'mcp-bridge-node',
          version: '1.0.0',
        },
      });

      if (response.result || response.error) {
        // Even "already initialized" means server is up
        this._connected = true;
        console.error(
          '[mcp-bridge] Connected to orchestrator ' +
            'via HTTP Streamable',
        );
        return true;
      }
      return false;
    } catch (err) {
      console.error('[mcp-bridge] Initialize failed:', err);
      this._connected = false;
      return false;
    }
  }

  async callTool(
    name: string,
    args: Record<string, unknown>,
  ): Promise<ToolCallResult> {
    const response = await this.post('tools/call', {
      name,
      arguments: args,
    });

    if (response.error) {
      return {
        content: [{
          type: 'text',
          text: JSON.stringify(response.error),
        }],
        isError: true,
      };
    }

    return (response.result as ToolCallResult) ?? {
      content: [{ type: 'text', text: '{}' }],
    };
  }

  async sendRequest(
    method: string,
    params?: Record<string, unknown>,
  ): Promise<JsonRpcResponse> {
    return this.post(method, params);
  }

  resetSession(): void {
    this._connected = false;
  }

  close(): void {
    this._connected = false;
  }

  private async post(
    method: string,
    params?: Record<string, unknown>,
  ): Promise<JsonRpcResponse> {
    const id = ++this.requestIdCounter;
    const body = JSON.stringify({
      jsonrpc: '2.0',
      id,
      method,
      ...(params && { params }),
    });

    const controller = new AbortController();
    const timeout = setTimeout(
      () => controller.abort(),
      this.config.requestTimeoutMs,
    );

    try {
      const res = await fetch(
        `${this.config.orchestratorUrl}/mcp`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body,
          signal: controller.signal,
        },
      );

      if (!res.ok) {
        throw new Error(
          `HTTP ${res.status}: ${res.statusText}`,
        );
      }

      const text = await res.text();
      if (!text) {
        throw new Error('Empty response');
      }
      return JSON.parse(text) as JsonRpcResponse;
    } finally {
      clearTimeout(timeout);
    }
  }
}

File: orchestrator-bridge/src/main/kotlin/com/orchestrator/mcp/bridge/HttpStreamableClient.kt
package com.orchestrator.mcp.bridge

import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.*
import org.slf4j.LoggerFactory
import java.util.UUID
import java.util.concurrent.atomic.AtomicLong

/**
 * HTTP Streamable client connecting to the MCP Orchestrator /mcp endpoint.
 * Manages session lifecycle and request/response handling.
 */
class HttpStreamableClient(private val config: BridgeConfig) {

    private val logger = LoggerFactory.getLogger(HttpStreamableClient::class.java)
    private val requestIdCounter = AtomicLong(0)
    private var sessionId: String? = null
    private var connected = false
    private var activeUrl: String = config.orchestratorUrl

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    private val httpClient = HttpClient(CIO) {
        install(ContentNegotiation) { json(json) }
        engine { requestTimeout = config.requestTimeoutMs }
    }

    val isConnected: Boolean get() = connected

    suspend fun initialize(url: String? = null, timeoutMs: Long? = null): Boolean {
        if (url != null) activeUrl = url
        val request = buildJsonRpcRequest("initialize", buildJsonObject {
            put("protocolVersion", JsonPrimitive("2025-03-26"))
            put("capabilities", buildJsonObject {})
            put("clientInfo", buildJsonObject {
                put("name", JsonPrimitive("mcp-bridge"))
                put("version", JsonPrimitive("1.0.0"))
            })
        })
        return try {
            val response = sendRawRequest(request, includeSession = false)
            sessionId = response.headers["Mcp-Session-Id"]
            connected = sessionId != null
            logger.info("Initialized session: $sessionId")
            connected
        } catch (e: Exception) {
            logger.error("Initialize failed: ${e.message}")
            connected = false
            false
        }
    }

    suspend fun sendRequest(method: String, params: JsonObject?): JsonObject {
        val request = buildJsonRpcRequest(method, params)
        val response = sendRawRequest(request, includeSession = true)
        return json.parseToJsonElement(response.bodyAsText()).jsonObject
    }

    /**
     * Call a tool via the MCP standard tools/call protocol.
     * Wraps the call correctly as:
     *   { "method": "tools/call", "params": { "name": toolName, "arguments": arguments } }
     * Then extracts and returns the concatenated text from result.content[].text.
     *
     * @throws Exception if the server returns a JSON-RPC error
     */
    suspend fun callTool(toolName: String, arguments: JsonObject?): String {
        val params = buildJsonObject {
            put("name", JsonPrimitive(toolName))
            arguments?.let { put("arguments", it) }
        }
        val request = buildJsonRpcRequest("tools/call", params)
        val response = sendRawRequest(request, includeSession = true)
        val jsonResponse = json.parseToJsonElement(response.bodyAsText()).jsonObject

        // Happy path: extract text from result.content[]
        val resultObj = jsonResponse["result"]?.jsonObject
        if (resultObj != null) {
            val contentArray = resultObj["content"]?.jsonArray
            if (!contentArray.isNullOrEmpty()) {
                return contentArray.joinToString("\n") { item ->
                    item.jsonObject["text"]?.jsonPrimitive?.content ?: ""
                }
            }
            // result exists but no content array — return raw result
            return resultObj.toString()
        }

        // Error path: propagate as exception so callers can return errorResult
        val errorObj = jsonResponse["error"]?.jsonObject
        if (errorObj != null) {
            val code = errorObj["code"]?.toString() ?: "UNKNOWN"
            val message = errorObj["message"]?.jsonPrimitive?.content ?: "Unknown error"
            throw Exception("tools/call '$toolName' failed [$code]: $message")
        }

        return "{}"
    }

    suspend fun close() {
        connected = false
        sessionId = null
        httpClient.close()
    }

    fun resetSession() {
        sessionId = null
        connected = false
        requestIdCounter.set(0)
    }

    private suspend fun sendRawRequest(body: String, includeSession: Boolean): HttpResponse {
        return httpClient.post("${activeUrl}/mcp") {
            contentType(ContentType.Application.Json)
            config.token?.let { header("Authorization", "Bearer $it") }
            if (includeSession && sessionId != null) {
                header("Mcp-Session-Id", sessionId)
            }
            setBody(body)
        }
    }

    private fun buildJsonRpcRequest(method: String, params: JsonObject?): String {
        val obj = buildJsonObject {
            put("jsonrpc", JsonPrimitive("2.0"))
            put("id", JsonPrimitive(requestIdCounter.incrementAndGet()))
            put("method", JsonPrimitive(method))
            params?.let { put("params", it) }
        }
        return json.encodeToString(JsonObject.serializer(), obj)
    }
}
